README.md
# json-ex

[![Build Status](https://travis-ci.org/gabrielcsapo/json-ex.svg?branch=master)](https://travis-ci.org/gabrielcsapo/json-ex)
[![Dependency Status](https://david-dm.org/gabrielcsapo/json-ex.svg)](https://david-dm.org/gabrielcsapo/json-ex)
[![devDependency Status](https://david-dm.org/gabrielcsapo/json-ex/dev-status.svg)](https://david-dm.org/gabrielcsapo/json-ex#info=devDependencies)
[![Coverage Status](https://lcov-server.herokuapp.com/badge/github%2Ecom/gabrielcsapo/json-ex.svg)](https://lcov-server.herokuapp.com/coverage/github%2Ecom/gabrielcsapo/json-ex)
[![npm](https://img.shields.io/npm/dt/json-ex.svg?maxAge=2592000)]()
[![npm](https://img.shields.io/npm/dm/json-ex.svg?maxAge=2592000)]()

> Extends JSON to  be able to serialize and deserialize more than just basic primitives

# What is this?

Let's parse JSON like it is 20**! Let's add support for functions, buffers, dates and the basic primitives! If you know of anymore let me know and let's add them!

> This is not a recreation of JSON, it is extending the protocol to allow for more types

# Install

```bash
npm install json-ex
```

## Usage

> json-ex is a drop in replacement for JSON, so it implements JSON.stringify and JSON.parse

### Stringify

```javascript
const JSONex = require('json-ex');
const object = {
    name: 'Hello world',
    person: true,
    age: 100000,
    buffer: new Buffer('hi'),
    date: new Date('10/20/2017'),
    func: function hello() { return 'hello world' },
    reg: new RegExp('%name%')
};
const output = JSONex.stringify(object);
```

#### output

```json
{
  "name": "Hello world",
  "person": true,
  "age": 100000,
  "buffer": "_BuffEx_%7B%22type%22%3A%22Buffer%22%2C%22data%22%3A%5B104%2C105%5D%7D",
  "date": "_DateEx_2017-10-20T07%3A00%3A00.000Z",
  "func": "_FuncRa_function%20hello()%20%7B%20return%20&#39;hello%20world&#39;%20%7D",
  "reg": "_PxEgEr_%5B%22%25name%25%22%2C%22%22%5D"
}
```

### Parse

```javascript
const string = `{
  "name": "Hello world",
  "person": true,
  "age": 100000,
  "buffer": "_BuffEx_%7B%22type%22%3A%22Buffer%22%2C%22data%22%3A%5B104%2C105%5D%7D",
  "date": "_DateEx_2017-10-20T07%3A00%3A00.000Z",
  "func": "_FuncRa_function%20hello()%20%7B%20return%20&#39;hello%20world&#39;%20%7D",
  "reg": "_PxEgEr_%5B%22%25name%25%22%2C%22%22%5D"
}`;
const output = JSONex.parse(string);
```

### output

```javascript
{ name: 'Hello world',
  person: true,
  age: 100000,
  buffer: <Buffer 68 69>,
  date: 2017-10-20T07:00:00.000Z,
  func: [Function: hello],
  reg: /%name%/
}
```

## Benchmark

> run benchmarks by running `npm run benchmark`

```bash
json-ex
  41,763 op/s » Stringify
 87,266 op/s » Parse


Suites:  1
Benches: 2
Elapsed: 1,760.74 ms
```


CHANGELOG.md
# 0.2.0 (09/28/2017)

- fixes an issue with serializing and deserializing regex
- update the distributable scripts name
- encodes string values instead of using JSON.stringify

# 0.1.0 (05/25/2017)

- fixes Stringify to actually return a string compliant string
- updates interface for parsing functions
- removes costly regex

# 0.0.1

- basic use cases covered


test/index.js
const test = require('tape');

const JSONex = require('../index');
const moment = require('moment');

test('json-ex', (t) => {
    t.plan(5);

    t.test('should be able to stringify a basic javascript object', (t) => {
        const object = {
            name: 'Hello world',
            person: true,
            age: 100000
        };
        const output = JSONex.stringify(object);
        t.deepEqual(output, '{"name":"Hello world","person":true,"age":100000}')
        t.end();
    });

    t.test('should be able to stringify a complex javascript object', (t) => {
        const object = {
            name: 'Hello world',
            person: true,
            age: 100000,
            buffer: new Buffer('hi'),
            date: new Date('10/20/2017'),
            func: function hello() { return 'hello world' },
            reg: new RegExp('%name%')
        };
        const output = JSONex.stringify(object);
        t.equal(output, '{"name":"Hello world","person":true,"age":100000,"buffer":"_BuffEx_%7B%22type%22%3A%22Buffer%22%2C%22data%22%3A%5B104%2C105%5D%7D","date":"_DateEx_2017-10-20T07%3A00%3A00.000Z","func":"_FuncRa_function%20hello()%20%7B%20return%20&#39;hello%20world&#39;%20%7D","reg":"_PxEgEr_%5B%22%25name%25%22%2C%22%22%5D"}');
        t.end();
    });

    t.test('should be able to parse a complex object', (t) => {
      const j = '{"name":"Hello world","person":true,"age":100000,"buffer":"_BuffEx_%7B%22type%22%3A%22Buffer%22%2C%22data%22%3A%5B104%2C105%5D%7D","date":"_DateEx_2017-10-20T07%3A00%3A00.000Z","func":"_FuncRa_function%20hello()%20%7B%20return%20&#39;hello%20world&#39;%20%7D","reg":"_PxEgEr_%5B%22%25name%25%22%2C%22%22%5D"}';
      const parsed = JSONex.parse(j);

      t.equal(parsed.date.constructor.name, 'Date');
      t.equal(parsed.buffer.constructor.name, 'Buffer');
      t.equal(parsed.reg.constructor.name, 'RegExp');
      t.equal(parsed.func.constructor.name, 'Function');

      t.end();
    });

    t.test('should be able to stringify a complex javascript object and parse it back', (t) => {
        const object = {
            name: 'Hello world',
            message: 'hi',
            person: true,
            age: 100000,
            buffer: new Buffer('hi'),
            date: new Date('10/20/2017'),
            func: function hello() { return 'hello world' },
            quickFunc: () => { return 'hello world' },
            reg: new RegExp('%name%'),
            moment
        };
        const output = JSONex.stringify(object);

        const parsed = JSONex.parse(output);

        t.equal(parsed.date.constructor.name, 'Date');
        t.equal(parsed.buffer.constructor.name, 'Buffer');
        t.equal(parsed.reg.constructor.name, 'RegExp');
        t.equal(parsed.quickFunc.constructor.name, 'Function');
        t.equal(parsed.func.constructor.name, 'Function');

        t.equal(parsed.func(), 'hello world');
        t.equal(parsed.quickFunc(), 'hello world');
        t.end();
    });

    t.test('should be able to stringify/parse RegExp with flags', (t) => {
        const object = {
          contrived: /example/i
        };

        const string = JSONex.stringify(object);
        const parsed = JSONex.parse(string);

        t.equal(object.contrived.test('ExAmPlE'), true);
        t.equal(parsed.contrived.test('ExAmPlE'), true);
        t.end();
    });
});


test/lib/util.js
const test = require('tape');

const { decode, encode } = require('../../lib/util');

test('@util', (t) => {
  t.plan(2);

  t.test('@encode', (t) => {
    t.plan(1);

    t.test('encode function', (t) => {
      t.equal(encode("function hello() { return 'hello world' }"), 'function%20hello()%20%7B%20return%20&#39;hello%20world&#39;%20%7D');
      t.end();
    });

  });

  t.test('@decode', (t) => {
    t.plan(1);

    t.test('encode function', (t) => {
      t.equal(decode("function%20hello()%20%7B%20return%20&#39;hello%20world&#39;%20%7D"), "function hello() { return 'hello world' }");
      t.end();
    });

  });

});


index.js
const { decode, encode } = require('./lib/util');

module.exports.stringify = function stringify(obj) {
  return JSON.stringify(obj, function(key, value) {
    var flags;

    if (obj[key] instanceof Function) {
      var rawBody = value.toString();
      var fnBody = encode(rawBody);
      if (rawBody.length < 8 || rawBody.substring(0, 8) !== 'function') { //this is ES6 Arrow Function
        return '_NuFrRa_' + fnBody;
      }

      return '_FuncRa_' + fnBody;
    }

    if (obj[key] instanceof Date) {
      return '_DateEx_' + encode(value);
    }
    if (value instanceof RegExp) {
      flags = value.flags;

      if (!flags) {
        // Older versions of node do not support the .flags property
        // and in such versions, we must recreate their value
        flags = '';

        if (value.global) {
          flags += 'g'
        }
        if (value.ignoreCase) {
          flags += 'i'
        }
        if (value.multiline) {
          flags += 'm'
        }
      }

      return encode([
        // To properly serialize a regular expression we need both
        // the .source and .flags properties.
        '_PxEgEr_["',
        // RegExp source escapes double quotes, for JSON we need
        // the opposite to be true
        value.source.replace(/\\'/gm, "'").replace(/"/gm, '\\"'),
        '","',
        // Flags do not have quotes in them so we are safe here
        flags,
        '"]'
      ].join(''));
    }
    // if it is an object check if that object has a class
    if (obj[key] instanceof Buffer) {
      return '_BuffEx_' + encode(JSON.stringify(obj[key]));
    }
    return value;
  });
}

module.exports.parse = function parse(str) {

  return JSON.parse(str, function(key, value) {
    var prefix;
    var array;
    var source;
    var flags;

    if (typeof value != 'string') {
      return value;
    }
    if (value.length < 8) {
      return value;
    }

    prefix = value.substring(0, 8);

    if (prefix === '_FuncRa_') {
      return eval('(' + decode(value.slice(8)) + ')');
    }
    if (prefix === '_NuFrRa_') {
      return eval(decode(value.slice(8)));
    }
    if (prefix === '_PxEgEr_') {
      array = JSON.parse(decode(value.slice(8)));
      source = array[0];
      flags = array[1];

      return new RegExp(source, flags);
    }
    if (prefix === '_DateEx_') {
      return new Date(decode(value.slice(8)));
    }
    if (prefix === '_BuffEx_') {
      return new Buffer(JSON.parse(decode(value.slice(8))));
    }

    return value;
  });
}


tryitout.js
module.exports = {
  title: "json-ex",
  description: "Parsing JSON like it is 2017",
  nav: {
    Source: "https://github.com/gabrielcsapo/json-ex"
  },
  options: {
    width: "90%"
  },
  body: [{
    type: "text",
    value: `
    <div style="text-align: center;">
      <blockquote style="width: auto;display: inline-block;">
        <p>
          <a href="https://github.com/gabrielcsapo/json-ex">
          <code>json-ex</code></a> extends JSON to be able to serialize and deserialize more than just basic primitives
        </p>
      </blockquote>
    </div>
    `
  }, {
    type: "code",
    title: "Encode",
    value: `
        var JSONex = require("json-ex");
        var output = JSONex.stringify({
            name: 'Hello world',
            person: true,
            age: 100000,
            date: new Date('10/20/2017'),
            func: function hello() { return 'hello world' },
            reg: new RegExp('%name%')
        });
        console.log(output);
      `
  }, {
    type: "code",
    title: "Decode",
    value: `
        var JSONex = require("json-ex");
        var output = JSONex.parse('{"name":"Hello world","person":true,"age":100000,"buffer":"_BuffEx_%7B%22type%22%3A%22Buffer%22%2C%22data%22%3A%5B104%2C105%5D%7D","date":"_DateEx_2017-10-20T07%3A00%3A00.000Z","func":"_FuncRa_function%20hello()%20%7B%20return%20&#39;hello%20world&#39;%20%7D","reg":"_PxEgEr_%5B%22%25name%25%22%2C%22%22%5D"}');
        console.log(output.func());
      `
  }],
  externals: [
    "./dist/jsonex.min.js"
  ],
  output: "./docs"
};
